### ISSUE TITLE
[PR #1270] fix(modkit): reject duplicate YAML mapping keys during config parsing

### ISSUE THREAD

**Issue by il10241024** [constructorfabric/gears-rust#3676]
> 🔗 **Mirrored PR** [cyberfabric/cyberware-rust#1270](https://github.com/cyberfabric/cyberware-rust/pull/1270) | **Author:** mattgarmon | **Opened:** 2026-03-20T21:00:34Z | **Status:** merged on 2026-03-20T22:58:30Z
> *GitHub API does not allow setting PR author or timestamps — attribution preserved here.*

---

Replace figment's built-in Yaml provider with a StrictYaml wrapper that uses serde_saphyr's DuplicateKeyPolicy::Error. This catches duplicate keys (e.g. two "module1:" entries under "modules:") at parse time instead of silently keeping the last value.

Fixes: https://github.com/constructorfabric/cyberfabric-core/issues/1255

## Summary by CodeRabbit

* **Bug Fixes**
  * Configuration loading now rejects duplicate YAML mapping keys (including within per-module files) and reports clear, file-specific errors instead of silently overriding earlier entries; module name duplicates are detected and surfaced.

* **Tests**
  * Added unit tests covering duplicate-key rejection, duplicate module name detection, and successful parsing of valid layered configurations.

---

**Comment by il10241024** [comment]
**coderabbitai[bot]** commented on 2026-03-20T21:00:52Z:

---


No actionable comments were generated in the recent review. 🎉

<details>
<summary>ℹ️ Recent review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `866790c0-5416-4cbe-8250-6258d986e81c`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between ed96e8de29772dfcc094f681875997f63262998c and aba9d854cd88388f7714fb2da3654b3d19f00115.

</details>

<details>
<summary>📒 Files selected for processing (1)</summary>

* `libs/modkit/src/bootstrap/config/mod.rs`

</details>

</details>

---




<details>
<summary>📝 Walkthrough</summary>

## Walkthrough

Replaced permissive YAML parsing with a strict parser that errors on duplicate mapping keys; wired a new `StrictYaml` figment provider and `strict_yaml_parse` helper into module-file merging, and added unit tests for duplicate-module and duplicate-key failures and success cases.

## Changes

|Cohort / File(s)|Summary|
|---|---|
|**Config parsing & tests** <br> `libs/modkit/src/bootstrap/config/mod.rs`|Introduced `StrictYaml` figment `Format` using `serde_saphyr` configured to error on duplicate mapping keys; added `strict_yaml_parse` helper; replaced `Yaml::file(...)` with `StrictYaml::file(...)` in `AppConfig::load_layered`; updated `merge_module_files` to parse module YAML with strict parsing and wrap errors with module filepath context; added unit tests for duplicate-module-name rejection, duplicate keys in module files, and successful unique-module parsing.|

## Sequence Diagram(s)

(omitted)

## Estimated code review effort

🎯 4 (Complex) | ⏱️ ~45 minutes

## Possibly related issues

- `#3003`: Modkit allows to have duplicate sections of same module config — This PR enforces erroring on duplicate module mapping keys, addressing the reported duplicate-section behavior.
- constructorfabric/cyberfabric-core#1271 — Implements strict YAML parsing via `serde_saphyr` to reject duplicate mapping keys, matching the approach described in that issue.

## Poem

> 🐇  
> I hop through configs, sniff and pry,  
> No mirrored keys will pass me by,  
> StrictYaml hushes duplicate claims,  
> Modules named once — tidy remains,  
> A carrot for clean files! 🥕

</details>






<details>
<summary>🚥 Pre-merge checks | ✅ 5</summary>

<details>
<summary>✅ Passed checks (5 passed)</summary>

|         Check name         | Status   | Explanation                                                                                                                                                                                 |
| :------------------------: | :------- | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
|      Description Check     | ✅ Passed | Check skipped - CodeRabbit’s high-level summary is enabled.                                                                                                                                 |
|         Title check        | ✅ Passed | The title clearly and concisely summarizes the main change: rejecting duplicate YAML mapping keys during config parsing, which directly addresses the changeset's core objective.           |
|     Linked Issues check    | ✅ Passed | The PR fully addresses issue `#3003` by implementing strict YAML parsing that rejects duplicate module names and duplicate keys, with tests validating both scenarios mentioned in the issue. |
| Out of Scope Changes check | ✅ Passed | All changes are in-scope: StrictYaml implementation, config loading updates, and unit tests directly support the objective of rejecting duplicate YAML keys.                                |
|     Docstring Coverage     | ✅ Passed | Docstring coverage is 100.00% which is sufficient. The required threshold is 80.00%.                                                                                                        |

</details>

<sub>✏️ Tip: You can configure your own custom pre-merge checks in the settings.</sub>

</details>





<details>
<summary>✨ Finishing Touches</summary>

<details>
<summary>🧪 Generate unit tests (beta)</summary>

- [ ]    Create PR with unit tests

</details>

</details>





---

Thanks for using [CodeRabbit](https://coderabbit.ai?utm_source=oss&utm_medium=github&utm_campaign=cyberfabric/cyberfabric-core&utm_content=1270)! It's free for OSS, and your support helps us grow. If you like it, consider giving us a shout-out.

<details>
<summary>❤️ Share</summary>

- [X](https://twitter.com/intent/tweet?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A&url=https%3A//coderabbit.ai)
- [Mastodon](https://mastodon.social/share?text=I%20just%20used%20%40coderabbitai%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20the%20proprietary%20code.%20Check%20it%20out%3A%20https%3A%2F%2Fcoderabbit.ai)
- [Reddit](https://www.reddit.com/submit?title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&text=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code.%20Check%20it%20out%3A%20https%3A//coderabbit.ai)
- [LinkedIn](https://www.linkedin.com/sharing/share-offsite/?url=https%3A%2F%2Fcoderabbit.ai&mini=true&title=Great%20tool%20for%20code%20review%20-%20CodeRabbit&summary=I%20just%20used%20CodeRabbit%20for%20my%20code%20review%2C%20and%20it%27s%20fantastic%21%20It%27s%20free%20for%20OSS%20and%20offers%20a%20free%20trial%20for%20proprietary%20code)

</details>

<sub>Comment `@coderabbitai help` to get the list of available commands and usage tips.</sub>

**Comment by il10241024** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-20T21:06:20Z:

---



<details>
<summary>🧹 Nitpick comments (1)</summary><blockquote>

<details>
<summary>libs/modkit/src/bootstrap/config/mod.rs (1)</summary><blockquote>

`306-306`: **Apply duplicate-key rejection consistently to `modules_dir` YAML files.**

Line 306 enforces strict duplicate-key handling for the main config, but module files are still parsed with permissive `serde_saphyr::from_str` in `merge_module_files` (Line 469). That leaves a second config path with different semantics.



<details>
<summary>Proposed consistency patch</summary>

```diff
diff --git a/libs/modkit/src/bootstrap/config/mod.rs b/libs/modkit/src/bootstrap/config/mod.rs
@@
-        let raw = fs::read_to_string(&path)?;
-        let json: serde_json::Value = serde_saphyr::from_str(&raw)?;
+        let raw = fs::read_to_string(&path)?;
+        let opts = serde_saphyr::Options {
+            duplicate_keys: serde_saphyr::DuplicateKeyPolicy::Error,
+            ..serde_saphyr::Options::default()
+        };
+        let json: serde_json::Value = serde_saphyr::from_str_with_options(&raw, opts)?;
         bag.insert(name, json);
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@libs/modkit/src/bootstrap/config/mod.rs` at line 306, The module YAML parsing
in merge_module_files currently uses permissive serde_saphyr::from_str which
allows duplicate keys, causing inconsistent semantics versus the main config
which uses StrictYaml::file; update merge_module_files to parse module files
with the same strict duplicate-key enforcement (e.g., use StrictYaml::str or
StrictYaml::file for each module content instead of serde_saphyr::from_str) so
module merges honor duplicate-key rejection; locate the merge_module_files
function and replace its serde_saphyr::from_str calls with the StrictYaml-based
parser used at the main config merge to ensure consistent behavior.
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Nitpick comments:
In `@libs/modkit/src/bootstrap/config/mod.rs`:
- Line 306: The module YAML parsing in merge_module_files currently uses
permissive serde_saphyr::from_str which allows duplicate keys, causing
inconsistent semantics versus the main config which uses StrictYaml::file;
update merge_module_files to parse module files with the same strict
duplicate-key enforcement (e.g., use StrictYaml::str or StrictYaml::file for
each module content instead of serde_saphyr::from_str) so module merges honor
duplicate-key rejection; locate the merge_module_files function and replace its
serde_saphyr::from_str calls with the StrictYaml-based parser used at the main
config merge to ensure consistent behavior.
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `d9e3a140-c606-4fde-9a09-3ef57a2c5dcd`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 8a0dbeca50a55c0414a584d0a303ba82e8c5ed9a and 105f34e0f7cfddd6a31edb00ae5676622523204b.

</details>

<details>
<summary>📒 Files selected for processing (1)</summary>

* `libs/modkit/src/bootstrap/config/mod.rs`

</details>

</details>

**Comment by il10241024** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-20T21:21:30Z:

---

**Actionable comments posted: 1**

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@libs/modkit/src/bootstrap/config/mod.rs`:
- Around line 419-438: The module file loader bypasses StrictYaml's
duplicate-key rejection because merge_module_files calls
serde_saphyr::from_str(&raw) directly; update merge_module_files to parse with
the same options used by StrictYaml (DuplicateKeyPolicy::Error) — e.g.,
construct serde_saphyr::Options with duplicate_keys set to
DuplicateKeyPolicy::Error and call serde_saphyr::from_str_with_options(raw,
opts) (or route parsing through StrictYaml::from_str) so module files are
validated the same way as the main config.
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `37ae11dd-a7b1-42ea-af31-bebdce7287cf`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 105f34e0f7cfddd6a31edb00ae5676622523204b and 6bc6f1342434668c8762674673deaa0d3c2cf774.

</details>

<details>
<summary>📒 Files selected for processing (1)</summary>

* `libs/modkit/src/bootstrap/config/mod.rs`

</details>

</details>

**Comment by il10241024** [comment]
**coderabbitai[bot]** submitted review **COMMENTED** on 2026-03-20T22:30:01Z:

---

**Actionable comments posted: 1**

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against the current code and only fix it if needed.

Inline comments:
In `@libs/modkit/src/bootstrap/config/mod.rs`:
- Around line 3078-3085: The YAML test string currently embeds modules_dir via
modules_dir.display(), which can produce backslashes on Windows and break
double-quoted YAML; normalize the path first (e.g., convert to a UTF-8 string
and replace backslashes with forward slashes or use to_string_lossy()) and then
use that normalized value in the cfg_yaml format call (identify the cfg_yaml
binding and the modules_dir.display() usage and replace with the normalized
modules_dir_str).
```

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro

**Run ID**: `e1df7569-2faa-4a29-bdd4-e936242daf6f`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 6bc6f1342434668c8762674673deaa0d3c2cf774 and ed96e8de29772dfcc094f681875997f63262998c.

</details>

<details>
<summary>📒 Files selected for processing (1)</summary>

* `libs/modkit/src/bootstrap/config/mod.rs`

</details>

</details>

**Comment by il10241024** [comment]
**coderabbitai[bot]** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 443 on 2026-03-20T21:21:29Z:

---

_⚠️ Potential issue_ | _🟠 Major_

<details>
<summary>🧩 Analysis chain</summary>

🏁 Script executed:

```shell
#!/bin/bash
rg -n -C2 'StrictYaml::file|serde_saphyr::from_str\(|serde_saphyr::from_str_with_options\(' libs/modkit/src/bootstrap/config/mod.rs
```

Repository: cyberfabric/cyberfabric-core

Length of output: 1290

---



</details>

**Apply duplicate-key rejection consistently to module file parsing.**

`StrictYaml` enforces duplicate-key rejection for the main config load path (Line 306), but `merge_module_files` at **Line 469** parses YAML using plain `serde_saphyr::from_str(&raw)` without the strict policy. This creates an inconsistency where module files bypass duplicate-key validation.

Apply the same `DuplicateKeyPolicy::Error` option to module file parsing:

<details>
<summary>Suggested fix</summary>

```diff
 fn merge_module_files(
     bag: &mut HashMap<String, serde_json::Value>,
     dir: impl AsRef<Path>,
 ) -> Result<()> {
@@
-        let raw = fs::read_to_string(&path)?;
-        let json: serde_json::Value = serde_saphyr::from_str(&raw)?;
+        let raw = fs::read_to_string(&path)?;
+        let opts = serde_saphyr::Options {
+            duplicate_keys: serde_saphyr::DuplicateKeyPolicy::Error,
+            ..serde_saphyr::Options::default()
+        };
+        let json: serde_json::Value = serde_saphyr::from_str_with_options(&raw, opts)?;
         bag.insert(name, json);
     }
     Ok(())
 }
```
</details>



<details>
<summary>📝 Committable suggestion</summary>

> ‼️ **IMPORTANT**
> Carefully review the code before committing. Ensure that it accurately replaces the highlighted code, contains no missing lines, and has no issues with indentation. Thoroughly test & benchmark the code to ensure it meets the requirements.

```suggestion
fn merge_module_files(
    bag: &mut HashMap<String, serde_json::Value>,
    dir: impl AsRef<Path>,
) -> Result<()> {
    let raw = fs::read_to_string(&path)?;
    let opts = serde_saphyr::Options {
        duplicate_keys: serde_saphyr::DuplicateKeyPolicy::Error,
        ..serde_saphyr::Options::default()
    };
    let json: serde_json::Value = serde_saphyr::from_str_with_options(&raw, opts)?;
    bag.insert(name, json);
    Ok(())
}
```

</details>



<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@libs/modkit/src/bootstrap/config/mod.rs` around lines 419 - 438, The module
file loader bypasses StrictYaml's duplicate-key rejection because
merge_module_files calls serde_saphyr::from_str(&raw) directly; update
merge_module_files to parse with the same options used by StrictYaml
(DuplicateKeyPolicy::Error) — e.g., construct serde_saphyr::Options with
duplicate_keys set to DuplicateKeyPolicy::Error and call
serde_saphyr::from_str_with_options(raw, opts) (or route parsing through
StrictYaml::from_str) so module files are validated the same way as the main
config.
```

</details>





✅ Confirmed as addressed by &#2369;mattgarmon

**Comment by il10241024** [comment]
**mattgarmon** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 443 on 2026-03-20T21:33:05Z:

---

Thanks, applied

**Comment by il10241024** [comment]
**coderabbitai[bot]** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 443 on 2026-03-20T21:33:35Z:

---

`@mattgarmon`, glad to hear it! The `merge_module_files` function will now enforce the same duplicate-key rejection as the main config load path, ensuring full consistency across all YAML parsing in modkit. 🎉




---

<details>
<summary>🧠 Learnings used</summary>

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 1175
File: libs/modkit/src/bootstrap/config/mod.rs:102-105
Timestamp: 2026-03-16T21:27:48.542Z
Learning: In `cyberfabric/cyberfabric-core`, the top-level `AppConfig` (in `libs/modkit/src/bootstrap/config/mod.rs`) is pre-release / under active development with no deployments yet as of PR `#1175` (March 2026). Breaking changes to config field names (e.g. renaming `tracing` to `opentelemetry`) do not require backward-compatibility shims such as `#[serde(alias = "tracing")]`. Do not flag config key renames in `AppConfig` as breaking changes until the project declares a stable/released state.
```

---

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 875
File: modules/mini-chat/mini-chat/src/config.rs:12-13
Timestamp: 2026-03-05T08:21:00.229Z
Learning: In `modules/mini-chat/mini-chat/src/config.rs`, the `vendor` field on `MiniChatConfig` is validated (trimmed and checked non-empty) inside `MiniChatModule::init` in `modules/mini-chat/mini-chat/src/module.rs`, not inside a `MiniChatConfig::validate()` method. Do not suggest adding a validate method to `MiniChatConfig` for the vendor field; the startup validation is already handled at the module init boundary.
```

---

```
Learnt from: mattgarmon
Repo: cyberfabric/cyberfabric-core PR: 1160
File: modules/system/oagw/oagw/src/domain/services/management.rs:570-625
Timestamp: 2026-03-16T21:50:45.202Z
Learning: In `modules/system/oagw/oagw/src/domain/services/management.rs`, the `validate_route_determinism` function performs a non-atomic read-before-write check (via `routes.list`) for route match conflicts. This is intentionally a temporary service-layer guard only; the authoritative uniqueness enforcement will be added at the repository/storage layer (e.g., DB unique constraint or transactional upsert) during the upcoming database integration work. Do not re-flag this service-layer check as a critical race condition bug — the team is aware and deferring the atomic fix to the database integration milestone.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 920
File: modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs:61-67
Timestamp: 2026-03-06T18:32:31.649Z
Learning: In `modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs`, silently skipping auth config when only one of `auth_plugin_type` or `auth_config` is `Some` is intentional. The existing `if let (Some(plugin_type), Some(config))` pattern that skips auth without a warning when there is a partial config is an accepted design decision. Do not flag the absence of a partial-auth-config warning as a bug.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 908
File: libs/modkit-db/src/outbox/sequencer.rs:118-127
Timestamp: 2026-03-09T11:24:39.573Z
Learning: In `libs/modkit-db/src/outbox/sequencer.rs`, the `claim_incoming_for_partition()` function's SQLite path performs a SELECT-then-DELETE without verifying the DELETE row count. This is intentional by design: SQLite is documented as single-process only (`Dialect::Sqlite` doc: "Do not run multiple outbox instances against the same SQLite database"), and SQLite's exclusive write locking within a transaction guarantees that no concurrent writer can claim or delete the same rows. Do not flag the missing `ExecResult`/`rows_affected` check in the SQLite claim path as a race condition or correctness issue.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 920
File: modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs:28-33
Timestamp: 2026-03-11T12:12:47.934Z
Learning: In `modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs`, the best-effort (warn-and-continue) provisioning pattern in `register_oagw_upstreams` is intentional for idempotent startup. OAGW's `create_upstream` returns an error for duplicate upstreams, so warn+skip is correct behavior on restart when upstreams/routes already exist. Fail-hard would prevent module restart without first cleaning up OAGW state. Genuine misconfigurations (non-duplicate failures) surface at runtime as `configuration_error` from `ProviderResolver`. Do not flag this warn-and-continue pattern as a bug or suggest making provisioning failures fatal.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1013
File: modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs:48-62
Timestamp: 2026-03-11T12:07:24.147Z
Learning: In `modules/mini-chat/mini-chat/src/infra/oagw_provisioning.rs`, the `anyhow::bail!` guard inside `register_oagw_upstreams` that fires when a tenant override has both `host: None` and `upstream_alias: None` is intentionally broader than the auth-only check in `ProviderEntry::validate()`. Any override missing both fields would produce a duplicate OAGW upstream alias (colliding with the parent), which is always invalid — not just for auth overrides. The design doc (Decision 3, override shape table) explicitly marks empty overrides as "Rejected by service: no distinct upstream possible". The two layers are complementary: `validate()` provides a user-friendly message for the auth-without-host shape; the service bail is defense-in-depth that catches ALL no-distinct-upstream shapes (including empty `{}`). Do not suggest narrowing this bail to auth-only overrides.
```

---

```
Learnt from: Artifizer
Repo: cyberfabric/cyberfabric-core PR: 949
File: dylint_lints/de09_gts_layer/de0901_gts_string_pattern/src/lib.rs:58-65
Timestamp: 2026-03-17T21:48:17.987Z
Learning: Global review guidance: The check_item function enforcing the _WILDCARD suffix convention for const/static GTS wildcard strings in dylint_lints/de09_gts_layer/de0901_gts_string_pattern/src/lib.rs should be treated as a cross-crate, cross-module convention and not limited to contract modules (i.e., do not scope it only to is_in_contract_module_ast). When reviewing similar code elsewhere, apply this naming convention check universally across all modules and crates unless there is an explicit, automated scope gate to restrict it.
```

---

```
Learnt from: striped-zebra-dev
Repo: cyberfabric/cyberfabric-core PR: 1162
File: modules/mini-chat/docs/DESIGN.md:397-397
Timestamp: 2026-03-17T09:49:31.465Z
Learning: In modules/mini-chat/docs/DESIGN.md, config knobs are documented using two parallel conventions: (1) Rust struct field names in snake_case (e.g., QuotaConfig.warning_threshold_pct) and (2) ConfigMap key paths with a section prefix (e.g., quota.warning_threshold_pct). This mirrors the pattern used for overshoot_tolerance_factor vs quota.overshoot_tolerance_factor and is intentional; do not flag these as inconsistent.
```

---

```
Learnt from: CR
Repo: cyberfabric/cyberfabric-core PR: 0
File: dylint_lints/AGENTS.md:0-0
Timestamp: 2026-02-09T12:27:51.227Z
Learning: Applies to dylint_lints/*/src/lib.rs : When module detection is needed, handle both `mod contract {}` inline modules and `contract/` directory-based modules
```

---

```
Learnt from: aviator5
Repo: cyberfabric/cyberfabric-core PR: 847
File: modules/mini-chat/mini-chat/src/domain/service/mod.rs:40-42
Timestamp: 2026-03-02T17:22:47.283Z
Learning: Enforce the GTS naming convention across Rust source files: use ~ as a delimiter where type definitions end with a trailing ~ (e.g., gts.x.core.users.user.v1~), instances separate type from instance with ~ (e.g., gts.acme.core.models.user.v1~acme.core.instances.user1.v1), and compound/aliased types chain multiple identifiers with ~ (e.g., gts.x.core.lic.feat.v1~x.core.global.base.v1). Do not flag a trailing or intermediate ~ as an error; treat it as intentional syntax. Apply these checks to Rust files (*.rs) throughout the codebase.
```

</details>

---



If you found this review helpful, would you consider giving us a shout-out on [X](https://twitter.com/intent/tweet?related=CodeRabbitAI&text=Enforced%20duplicate-key%20rejection%20for%20module%20YAML%20parsing%3A%20use%20serde_saphyr%3A%3Afrom_str_with_options%20%2B%20DuplicateKeyPolicy%3A%3AError%20so%20module%20files%20match%20main%20config%20validation.%20Try%20%40coderabbitai%20%F0%9F%9A%80%F0%9F%94%92&url=https%3A%2F%2Fgithub.com%2Fcyberfabric%2Fcyberfabric-core%2Fpull%2F1270%23discussion_r2968035577)?

Thank you for using CodeRabbit!

**Comment by il10241024** [comment]
**MikeFalcon77** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 3038 on 2026-03-20T22:12:15Z:

---

**MEDIUM**

`merge_module_files` now uses `strict_yaml_parse`, but there is no test verifying that a module config file with duplicate keys is rejected.

Both changed code paths must be tested: the `load_layered` path is covered, but a module file with a duplicate key would be parsed via `merge_module_files` and that rejection is untested.

Add a test that writes a module YAML file with a duplicate key into a temp modules dir and asserts that the load path which calls `merge_module_files` returns an error.

**Comment by il10241024** [comment]
**mattgarmon** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 3038 on 2026-03-20T22:18:09Z:

---

Added the missing test

**Comment by il10241024** [comment]
**coderabbitai[bot]** reviewed `libs/modkit/src/bootstrap/config/mod.rs` line 3085 on 2026-03-20T22:30:00Z:

---

_⚠️ Potential issue_ | _🟡 Minor_

**Normalize `modules_dir` before embedding it in YAML test content.**

Using `modules_dir.display()` inside a double-quoted YAML string can break on Windows due to backslash escape handling, causing failures unrelated to duplicate-key behavior.

<details>
<summary>Suggested fix</summary>

```diff
-        let cfg_yaml = format!(
+        let modules_dir_str = modules_dir.to_string_lossy().replace('\\', "/");
+        let cfg_yaml = format!(
             r#"
 server:
   home_dir: "~/.test_dup_modfile"
 modules_dir: "{}"
 "#,
-            modules_dir.display()
+            modules_dir_str
         );
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@libs/modkit/src/bootstrap/config/mod.rs` around lines 3078 - 3085, The YAML
test string currently embeds modules_dir via modules_dir.display(), which can
produce backslashes on Windows and break double-quoted YAML; normalize the path
first (e.g., convert to a UTF-8 string and replace backslashes with forward
slashes or use to_string_lossy()) and then use that normalized value in the
cfg_yaml format call (identify the cfg_yaml binding and the
modules_dir.display() usage and replace with the normalized modules_dir_str).
```

</details>
